// @droneblocks/go1-js/dist/mqtt/go1-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGo1StateCopy = void 0;
const data = {
    mqttConnected: false,
    managerOn: false,
    controllerOn: false,
    bms: {
        version: "unknown",
        status: 0,
        soc: 0,
        current: 0,
        cycle: 0,
        temps: new Array(4).fill(0),
        voltage: 0,
        cellVoltages: new Array(10).fill(0),
    },
    robot: {
        version: {
            hardware: "--",
            software: "--",
        },
        sn: {
            product: "--",
            id: "--",
        },
        temps: new Array(20).fill(0),
        mode: 0,
        gaitType: 0,
        obstacles: [255, 255, 255, 255],
        state: "invalid",
        distanceWarning: {
            front: 0,
            back: 0,
            left: 0,
            right: 0,
        },
    },
};
const dataCopy = JSON.stringify(data);
const getGo1StateCopy = () => {
    return JSON.parse(dataCopy);
};
exports.getGo1StateCopy = getGo1StateCopy;
