// @droneblocks/go1-js/dist/mqtt/go1-mqtt.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Go1MQTT = void 0;
const mqtt_1 = require("mqtt");
const go1_state_1 = require("./go1-state");
const message_handler_1 = require("./message-handler");
class Go1MQTT {
    constructor(go1, iClientOptions) {
        this.floats = new Float32Array(4);
        this.connected = false;
        this.defaultIClientOptions = {
            port: 1883,
            host: "192.168.12.1",
            clientId: Math.random().toString(16).substring(2, 8),
            keepalive: 5,
            protocol: "mqtt",
        };
        this.connect = () => {
            console.log("connecting");
            this.client = (0, mqtt_1.connect)(this.iClientOptions);
            this.client.on("connect", () => {
                console.log("connected");
                this.connected = true;
                this.go1.publishConnectionStatus(true);
            });
            this.client.on("close", () => {
                this.connected = false;
            });
            this.client.on("disconnect", () => {
                console.log("disconnected");
                this.go1.publishConnectionStatus(false);
            });
            this.client.on("offline", () => {
                console.log("disconnected");
                this.go1.publishConnectionStatus(false);
            });
            // Handle messages that come from various topics
            this.client.on("message", (topic, message) => {
                this.go1.publishState(this.go1State);
                (0, message_handler_1.default)(topic, message, this.go1State);
            });
        };
        // Subscribe to topics for updates
        this.subscribe = () => {
            var _a;
            (_a = this.client) === null || _a === void 0 ? void 0 : _a.subscribe(["bms/state", "firmware/version"]);
        };
        this.getState = () => {
            return this.go1State;
        };
        this.disconnect = () => {
            var _a;
            (_a = this.client) === null || _a === void 0 ? void 0 : _a.end();
        };
        this.updateSpeed = (leftRight, turnLeftRight, lookUpDown, // Only for stand mode
        backwardForward) => {
            this.floats[0] = this.clamp(leftRight);
            this.floats[1] = this.clamp(turnLeftRight);
            this.floats[2] = this.clamp(lookUpDown);
            this.floats[3] = this.clamp(backwardForward);
        };
        this.sendMovementCommand = (lengthOfTime) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            let interval;
            let zero = new Float32Array(4);
            zero[0] = 0;
            zero[1] = 0;
            zero[2] = 0;
            zero[3] = 0;
            // Reset speed from the buffer for a few
            (_a = this.client) === null || _a === void 0 ? void 0 : _a.publish(this.movementTopic, new Uint8Array(zero.buffer), {
                qos: 0,
            });
            interval = setInterval(() => {
                var _a;
                console.log(`sending command ${this.floats}`);
                (_a = this.client) === null || _a === void 0 ? void 0 : _a.publish(this.movementTopic, new Uint8Array(this.floats.buffer), {
                    qos: 0,
                });
            }, this.publishFrequency);
            // Stop sending after lengthOfTime
            return new Promise((resolve) => {
                setTimeout(() => {
                    resolve();
                    clearInterval(interval);
                }, lengthOfTime);
            });
        });
        this.sendLEDCommand = (r, g, b) => {
            var _a;
            (_a = this.client) === null || _a === void 0 ? void 0 : _a.publish(this.ledTopic, `child_conn.send('change_light(${r},${g},${b})')`, {
                qos: 0,
            });
        };
        this.sendModeCommand = (mode) => {
            var _a;
            (_a = this.client) === null || _a === void 0 ? void 0 : _a.publish(this.modeTopic, mode, {
                qos: 1,
            });
        };
        this.clamp = (speed) => {
            if (speed < -1.0) {
                return -1.0;
            }
            else if (speed > 1.0) {
                return 1.0;
            }
            else {
                return speed;
            }
        };
        this.go1 = go1;
        this.iClientOptions = Object.assign(Object.assign({}, this.defaultIClientOptions), iClientOptions);
        this.client = null;
        this.floats[0] = 0; // walk left (neg) and right (pos)
        this.floats[1] = 0; // turn left (neg) and  right (pos)
        this.floats[2] = 0;
        this.floats[3] = 0; // walk backward (neg) and forward (pos)
        this.movementTopic = "controller/stick";
        this.ledTopic = "programming/code";
        this.modeTopic = "controller/action";
        this.publishFrequency = 100; // Send MQTT message every 100ms
        this.go1State = (0, go1_state_1.getGo1StateCopy)();
    }
}
exports.Go1MQTT = Go1MQTT;
